lib/benchmark/benchmark.go

```diff
@@ -149,157 +158,113 @@ func (c *Config) Benchmark(ctx context.Context, tc *client.TeleportClient) (Resu
 		}
-	}
-}
-
-func produceMeasures(ctx context.Context, rate int, c chan<- benchMeasure) {
-	interval := time.Duration(1 / float64(rate) * float64(time.Second))
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ticker.C:
 
-			measure := benchMeasure{
-				Start: time.Now(),
-			}
-			select {
-			case c <- measure:
-			case <-ctx.Done():
-				return
-			}
-		case <-ctx.Done():
-			return
-		}
 	}
 }
 
 type benchMeasure struct {
-	Start           time.Time
-	End             time.Time
-	ThreadCompleted bool
-	ThreadID        int
-	Error           error
+	ResponseStart time.Time
+	End           time.Time
+	Error         error
+	client        *client.TeleportClient
+	command       []string
+	interactive   bool
 }
 
-type benchmarkThread struct {
-	id          int
-	ctx         context.Context
-	client      *client.TeleportClient
-	command     []string
-	interactive bool
-	receiveC    chan benchMeasure
-	sendC       chan benchMeasure
+func work(ctx context.Context, m benchMeasure, send chan<- benchMeasure) {
+	m.Error = execute(m)
+	m.End = time.Now()
+	select {
+	case send <- m:
+	case <-ctx.Done():
+		return
+	}
 }
 
-func (b *benchmarkThread) execute(measure benchMeasure) {
-	if !b.interactive {
+func execute(m benchMeasure) error {
+	if !m.interactive {
 		// do not use parent context that will cancel in flight requests
 		// because we give test some time to gracefully wrap up
 		// the in-flight connections to avoid extra errors
-		measure.Error = b.client.SSH(context.TODO(), nil, false)
-		measure.End = time.Now()
-		b.sendMeasure(measure)
-		return
+		return m.client.SSH(context.TODO(), m.command, false)
 	}
-	config := b.client.Config
+	config := m.client.Config
 	client, err := client.NewClient(&config)
 	if err != nil {
-		measure.Error = err
-		measure.End = time.Now()
-		b.sendMeasure(measure)
-		return
+		return err
 	}
 	reader, writer := io.Pipe()
 	defer reader.Close()
 	defer writer.Close()
 	client.Stdin = reader
-	out := &bytes.Buffer{}
+	out := &utils.SyncBuffer{}
 	client.Stdout = out
 	client.Stderr = out
-	done := make(chan bool)
-	go func() {
-		measure.Error = b.client.SSH(context.TODO(), nil, false)
-		measure.End = time.Now()
-		b.sendMeasure(measure)
-		close(done)
-	}()
-	writer.Write([]byte(strings.Join(b.command, " ") + "\r\nexit\r\n"))
-	<-done
-}
-
-func (b *benchmarkThread) sendMeasure(measure benchMeasure) {
-	measure.ThreadID = b.id
-	select {
-	case b.sendC <- measure:
-	default:
-		logrus.Warning("blocked on measure send")
-	}
-}
-
-func (b *benchmarkThread) run() {
-	for {
-		select {
-		case measure := <-b.receiveC:
-			b.execute(measure)
-		case <-b.ctx.Done():
-			b.sendMeasure(benchMeasure{
-				ThreadCompleted: true,
-			})
-			return
-		}
+	err = m.client.SSH(context.TODO(), nil, false)
+	if err != nil {
+		return err
 	}
+	writer.Write([]byte(strings.Join(m.command, " ") + "\r\nexit\r\n"))
```

Comment:
Maybe for a future PR: check the error from `Write`, to make sure we actually successfully write the command out.

Also, I start to suspect that we're not actually running an interactive shell command at all.
`m.client.SSH` looks like it should block until the interactive session is over. So we don't get to write the command at all - session is over once `m.client.SSH` returns.

If I'm right, we should replace the `io.Pipe` with a simple `bytes.Reader` containing the command.
If I'm wrong, it's unclear when the SSH session ends, since we're not deferring any cleanup code matching `m.client.SSH`.